### PR TITLE
Require data permissions to set parameter mappings

### PR DESCRIPTION
**Pull request by camsaul**
Part of https://github.com/metabase/metaboat/issues/144

Users now have to have appropriate _data_ permissions to add or modify parameter mappings for DashboardCards. e.g. you must have _data_ permissions for the `venues` Table in order to map a Dashboard parameter to the Field `venues.category_id`. 

`PUT /api/dashboard/:id/cards` only checks the permissions for parameter mappings you're adding or changing. This means you can still reposition and resize cards and things like that without having to have data permissions.

- [x] Add validation to `POST /api/dashboard/:id/cards` (add dashboard cards endpoint)
- [x] Add validation to `PUT /api/dashboard/:id/cards` (modify dashboard cards endpoint)
- [x] Tests for `POST /api/dashboard/:id/cards` 
- [x] Tests for `PUT /api/dashboard/:id/cards` 
- [x] Tests for EE sandboxed permissions

### CODE REVIEW

File: src/metabase/api/dashboard.clj [hunk 1]
```diff
@@ -336,24 +379,78 @@
...
+              :sizeX              ...
+              :sizeY              ...
+              :row                ...
+              :col                ...
+              :parameter_mappings ...
+              :series             [{:id 123
+                                    ...}]}
+             ...]}"
   [id :as {{:keys [cards]} :body}]
+  {cards (su/non-empty [UpdatedDashboardCard])}
```
Comment (camsaul) [inline src/metabase/api/dashboard.clj:451]: Not specifically related to this change but I added a schema for validation and better dox to make it clearer what one is expected to pass in here

File: src/metabase/models/dashboard.clj [hunk 2]
```diff
@@ -291,8 +291,11 @@
         dashcard-ids        (db/select-ids DashboardCard, :dashboard_id (u/the-id dashboard-or-id))]
     (doseq [{dashcard-id :id, :as dashboard-card} dashcards]
       ;; ensure the dashcard we are updating is part of the given dashboard
-      (when (contains? dashcard-ids dashcard-id)
-        (dashboard-card/update-dashboard-card! (update dashboard-card :series #(filter identity (map :id %))))))
+      (if (contains? dashcard-ids dashcard-id)
+        (dashboard-card/update-dashboard-card! (update dashboard-card :series #(filter identity (map :id %))))
+        (throw (ex-info (tru "Dashboard {0} does not have a DashboardCard with ID {1}"
+                             (u/the-id dashboard-or-id) dashcard-id)
+                        {:status-code 404}))))
```
Comment (camsaul) [inline src/metabase/models/dashboard.clj:298]: Another change not specifically related to this PR. I changed things so instead of silently ignoring DashboardCards you attempt to update that don't exist, it will now return a 404

File: src/metabase/models/dashboard_card.clj [hunk 3]
```diff
@@ -35,11 +36,21 @@
...
+(defn normalize-parameter-mappings
+  "Normalize `parameter-mappings` when coming out of the application database or in via an API request."
+  [parameter-mappings]
+  (or (normalize/normalize-fragment [:parameters] parameter-mappings)
+      []))
+
+(models/add-type! ::parameter-mappings
+  :in  (comp i/json-in normalize-parameter-mappings)
+  :out (comp (i/catch-normalization-exceptions normalize-parameter-mappings) i/json-out-with-keywordization))
+
```
Comment (camsaul) [inline src/metabase/models/dashboard_card.clj:48]: One more tangentially related change. I moved the `:parameter-mappings` Toucan column type definition into the DashboardCard namespace since that's the only model that uses it and I needed to use functions powering it elsewhere

File: test/metabase/test/util/log.clj [hunk 4]
```diff
@@ -166,11 +166,14 @@
...
-    (.setLevel logger new-level)
-    (.updateLoggers (logger-context))))
+  ([new-level :- LogLevelKeyword]
+   (set-ns-log-level! (ns-name *ns*) new-level))
+
+  ([ns-symb new-level :- LogLevelKeyword]
+   (let [logger    (get-or-create-ns-logger! ns-symb)
+         new-level (->Level new-level)]
+     (.setLevel logger new-level)
+     (.updateLoggers (logger-context)))))
```
Comment (camsaul) [inline test/metabase/test/util/log.clj:176]: Actually, one more unrelated change. I tweaked this REPL util function to add a 1-arity to change the log level for the current namespace

**Review by dpsutton — APPROVED** [review]
